File: app/src/main/java/com/cameronbanga/skyscraper/viewmodels/PostComposerViewModel.kt
package com.cameronbanga.skyscraper.viewmodels

import android.app.Application
import android.content.ContentResolver
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.cameronbanga.skyscraper.models.*
import com.cameronbanga.skyscraper.services.ATProtoClient
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel for post composition
 * Handles text, images, language, moderation settings, mentions, hashtags, and posting
 */
class PostComposerViewModel(
    application: Application,
    private val replyTo: ReplyRef? = null,
    private val draft: PostDraft? = null
) : AndroidViewModel(application) {

    private val client = ATProtoClient.getInstance()
    private val contentResolver: ContentResolver = application.contentResolver

    private val _text = MutableStateFlow("")
    val text: StateFlow<String> = _text.asStateFlow()

    private val _selectedImages = MutableStateFlow<List<Uri>>(emptyList())
    val selectedImages: StateFlow<List<Uri>> = _selectedImages.asStateFlow()

    private val _imageAltTexts = MutableStateFlow<List<String>>(emptyList())
    val imageAltTexts: StateFlow<List<String>> = _imageAltTexts.asStateFlow()

    private val _selectedLanguage = MutableStateFlow(Language.defaultLanguage)
    val selectedLanguage: StateFlow<Language> = _selectedLanguage.asStateFlow()

    private val _moderationSettings = MutableStateFlow(ModerationSettings.default)
    val moderationSettings: StateFlow<ModerationSettings> = _moderationSettings.asStateFlow()

    private val _isPosting = MutableStateFlow(false)
    val isPosting: StateFlow<Boolean> = _isPosting.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _mentionSuggestions = MutableStateFlow<List<Profile>>(emptyList())
    val mentionSuggestions: StateFlow<List<Profile>> = _mentionSuggestions.asStateFlow()

    private val _hashtagSuggestions = MutableStateFlow<List<String>>(emptyList())
    val hashtagSuggestions: StateFlow<List<String>> = _hashtagSuggestions.asStateFlow()

    private val _cursorPosition = MutableStateFlow(0)
    private val cursorPosition: StateFlow<Int> = _cursorPosition.asStateFlow()

    private var mentionSearchJob: Job? = null
    private var hashtagSearchJob: Job? = null
    private var currentDraftId: String? = draft?.id

    val characterCount: StateFlow<Int> = MutableStateFlow(0).apply {
        viewModelScope.launch {
            text.collect { value = it.length }
        }
    }

    val canPost: StateFlow<Boolean> = MutableStateFlow(false).apply {
        viewModelScope.launch {
            text.collect {
                val hasText = it.trim().isNotEmpty()
                val hasMedia = _selectedImages.value.isNotEmpty()
                val withinLimit = it.length <= 300
                value = (hasText || hasMedia) && withinLimit
            }
        }
    }

    val canSaveDraft: StateFlow<Boolean> = MutableStateFlow(false).apply {
        viewModelScope.launch {
            text.collect {
                value = it.trim().isNotEmpty() || _selectedImages.value.isNotEmpty()
            }
        }
    }

    init {
        // Load draft if provided
        draft?.let { loadDraft(it) }
    }

    private fun loadDraft(draft: PostDraft) {
        _text.value = draft.text
        _imageAltTexts.value = draft.imageAltTexts
        _selectedLanguage.value = Language.allLanguages.firstOrNull { it.id == draft.languageId }
            ?: Language.defaultLanguage
        _moderationSettings.value = draft.moderationSettings
        // Note: selectedImages would need to be restored from imageData if needed
    }

    fun updateText(newText: String) {
        _text.value = newText
        _cursorPosition.value = newText.length
        detectMention()
        detectHashtag()
    }

    fun loadSelectedImages(uris: List<Uri>) {
        _selectedImages.value = uris
        // Initialize alt texts for new images
        _imageAltTexts.value = List(uris.size) { index ->
            _imageAltTexts.value.getOrNull(index) ?: ""
        }
    }

    fun removeImage(index: Int) {
        if (index < _selectedImages.value.size) {
            _selectedImages.value = _selectedImages.value.filterIndexed { i, _ -> i != index }
            _imageAltTexts.value = _imageAltTexts.value.filterIndexed { i, _ -> i != index }
        }
    }

    fun updateAltText(index: Int, altText: String) {
        if (index < _imageAltTexts.value.size) {
            _imageAltTexts.value = _imageAltTexts.value.toMutableList().apply {
                this[index] = altText
            }
        }
    }

    fun setLanguage(language: Language) {
        _selectedLanguage.value = language
    }

    fun setModerationSettings(settings: ModerationSettings) {
        _moderationSettings.value = settings
    }

    // MARK: - Mention Autocomplete

    private fun detectMention() {
        mentionSearchJob?.cancel()

        val cursorIndex = _cursorPosition.value.coerceIn(0, _text.value.length)
        val textBeforeCursor = _text.value.substring(0, cursorIndex)

        // Find the last @ symbol before cursor
        val lastAtIndex = textBeforeCursor.lastIndexOf('@')
        if (lastAtIndex >= 0) {
            val mentionText = textBeforeCursor.substring(lastAtIndex + 1)

            // Check if there's a space (which would end the mention)
            if (!mentionText.contains(' ')) {
                searchMentions(mentionText)
            } else {
                clearMentionSuggestions()
            }
        } else {
            clearMentionSuggestions()
        }
    }

    private fun searchMentions(query: String) {
        mentionSearchJob?.cancel()

        mentionSearchJob = viewModelScope.launch {
            try {
                // Add delay to avoid too many API calls
                delay(300)

                val results = mutableListOf<Profile>()
                val seenDIDs = mutableSetOf<String>()

                // Get current user's DID
                val currentUserDID = client.session.value?.did ?: return@launch

                // 1. Search follows
                try {
                    val follows = client.getFollows(currentUserDID, limit = 50)
                    follows.forEach { profile ->
                        if (query.isEmpty() ||
                            profile.handle.contains(query, ignoreCase = true) ||
                            (profile.displayName?.contains(query, ignoreCase = true) == true)
                        ) {
                            if (!seenDIDs.contains(profile.did)) {
                                results.add(profile)
                                seenDIDs.add(profile.did)
                            }
                        }
                    }
                } catch (e: Exception) {
                    // Continue even if follows search fails
                }

                // 2. Search followers
                try {
                    val followers = client.getFollowers(currentUserDID, limit = 50)
                    followers.forEach { profile ->
                        if (query.isEmpty() ||
                            profile.handle.contains(query, ignoreCase = true) ||
                            (profile.displayName?.contains(query, ignoreCase = true) == true)
                        ) {
                            if (!seenDIDs.contains(profile.did)) {
                                results.add(profile)
                                seenDIDs.add(profile.did)
                            }
                        }
                    }
                } catch (e: Exception) {
                    // Continue even if followers search fails
                }

                // 3. Search all users if we need more results
                if (results.size < 10 && query.isNotEmpty()) {
                    try {
                        val searchResults = client.searchUsers(query, limit = 10)
                        searchResults.forEach { profile ->
                            if (!seenDIDs.contains(profile.did)) {
                                results.add(profile)
                                seenDIDs.add(profile.did)
                            }
                        }
                    } catch (e: Exception) {
                        // Continue even if search fails
                    }
                }

                // Limit to 10 results
                _mentionSuggestions.value = results.take(10)
            } catch (e: Exception) {
                // Ignore errors in mention search
            }
        }
    }

    fun insertMention(profile: Profile) {
        val cursorIndex = _cursorPosition.value.coerceIn(0, _text.value.length)
        val textBeforeCursor = _text.value.substring(0, cursorIndex)

        // Find the last @ symbol
        val lastAtIndex = textBeforeCursor.lastIndexOf('@')
        if (lastAtIndex >= 0) {
            val beforeAt = _text.value.substring(0, lastAtIndex)
            val afterCursor = _text.value.substring(cursorIndex)

            // Insert the mention
            val mention = "@${profile.handle} "
            _text.value = beforeAt + mention + afterCursor
            _cursorPosition.value = beforeAt.length + mention.length

            clearMentionSuggestions()
        }
    }

    private fun clearMentionSuggestions() {
        _mentionSuggestions.value = emptyList()
        mentionSearchJob?.cancel()
    }

    // MARK: - Hashtag Autocomplete

    private fun detectHashtag() {
        hashtagSearchJob?.cancel()

        val cursorIndex = _cursorPosition.value.coerceIn(0, _text.value.length)
        val textBeforeCursor = _text.value.substring(0, cursorIndex)

        // Find the last # symbol before cursor
        val lastHashIndex = textBeforeCursor.lastIndexOf('#')
        if (lastHashIndex >= 0) {
            val hashtagText = textBeforeCursor.substring(lastHashIndex + 1)

            // Check if there's a space (which would end the hashtag)
            if (!hashtagText.contains(' ')) {
                searchHashtags(hashtagText)
            } else {
                clearHashtagSuggestions()
            }
        } else {
            clearHashtagSuggestions()
        }
    }

    private fun searchHashtags(query: String) {
        hashtagSearchJob?.cancel()

        hashtagSearchJob = viewModelScope.launch {
            try {
                // Add delay to avoid too many API calls
                delay(300)

                // Search for posts with this hashtag
                val searchQuery = if (query.isEmpty()) "popular" else "#$query"
                val response = client.searchPosts(searchQuery, limit = 20)

                // Extract unique hashtags from posts
                val hashtags = mutableSetOf<String>()
                response.posts.forEach { post ->
                    post.record.tags?.forEach { tag ->
                        hashtags.add(tag)
                    }
                }

                // Filter by query and limit to 10
                val filteredHashtags = hashtags.sorted().filter { tag ->
                    query.isEmpty() || tag.startsWith(query, ignoreCase = true)
                }.take(10)

                _hashtagSuggestions.value = filteredHashtags
            } catch (e: Exception) {
                // Ignore errors in hashtag search
            }
        }
    }

    fun insertHashtag(tag: String) {
        val cursorIndex = _cursorPosition.value.coerceIn(0, _text.value.length)
        val textBeforeCursor = _text.value.substring(0, cursorIndex)

        // Find the last # symbol
        val lastHashIndex = textBeforeCursor.lastIndexOf('#')
        if (lastHashIndex >= 0) {
            val beforeHash = _text.value.substring(0, lastHashIndex)
            val afterCursor = _text.value.substring(cursorIndex)

            // Insert the hashtag
            val hashtag = "#$tag "
            _text.value = beforeHash + hashtag + afterCursor
            _cursorPosition.value = beforeHash.length + hashtag.length

            clearHashtagSuggestions()
        }
    }

    private fun clearHashtagSuggestions() {
        _hashtagSuggestions.value = emptyList()
        hashtagSearchJob?.cancel()
    }

    // MARK: - Posting

    fun post(onComplete: (Boolean) -> Unit) {
        if (!canPost.value) {
            onComplete(false)
            return
        }

        viewModelScope.launch {
            _isPosting.value = true
            _errorMessage.value = null

            try {
                // Upload images if any are selected
                val uploadedImages = mutableListOf<UploadedImage>()
                _selectedImages.value.forEachIndexed { index, uri ->
                    try {
                        val inputStream = contentResolver.openInputStream(uri)
                        val bytes = inputStream?.readBytes()
                        inputStream?.close()

                        if (bytes != null) {
                            val altText = _imageAltTexts.value.getOrNull(index)
                                ?.takeIf { it.isNotEmpty() }
                            val uploaded = client.uploadImageBytes(bytes, altText)
                            uploadedImages.add(uploaded)
                        }
                    } catch (e: Exception) {
                        // Continue even if one image fails
                    }
                }

                // Create post
                client.createPost(
                    text = _text.value,
                    reply = replyTo,
                    images = uploadedImages.ifEmpty { null },
                    langs = listOf(_selectedLanguage.value.id)
                )

                // Clear form on success
                _text.value = ""
                _selectedImages.value = emptyList()
                _imageAltTexts.value = emptyList()

                onComplete(true)
            } catch (e: Exception) {
                _errorMessage.value = e.message ?: "Failed to create post"
                onComplete(false)
            } finally {
                _isPosting.value = false
            }
        }
    }

    fun saveDraft() {
        if (!canSaveDraft.value) return

        // TODO: Implement draft saving to local storage
        // This would require a DraftManager similar to the iOS version
    }

    fun clearError() {
        _errorMessage.value = null
    }
}

File: Skyscraper-iOS-20251106/Skyscraper/ViewModels/PostComposerViewModel.swift
//
//  PostComposerViewModel.swift
//  Skyscraper
//
//  Handles post composition and submission
//

import Foundation
import SwiftUI
import Combine
import PhotosUI

#if os(iOS)
import UIKit
#elseif os(macOS)
import AppKit
#endif

@MainActor
class PostComposerViewModel: ObservableObject {
    @Published var text = "" {
        didSet {
            detectMention()
            detectHashtag()
        }
    }
    @Published var isPosting = false
    @Published var errorMessage: String?
    @Published var selectedPhotoItems: [PhotosPickerItem] = []
    @Published var selectedImages: [PlatformImage] = []
    @Published var imageAltTexts: [String] = [] // Alt text for each image
    @Published var mentionSuggestions: [Profile] = []
    @Published var isSearchingMentions = false
    @Published var currentMentionQuery: String?
    @Published var hashtagSuggestions: [String] = []
    @Published var isSearchingHashtags = false
    @Published var currentHashtagQuery: String?
    @Published var cursorPosition: Int = 0
    @Published var selectedLanguage: Language
    @Published var moderationSettings: PostModerationSettings

    private let client = ATProtoClient.shared
    private let replyTo: ReplyRef?
    private var mentionSearchTask: Task<Void, Never>?
    private var hashtagSearchTask: Task<Void, Never>?
    private var currentDraftId: UUID?

    var characterCount: Int {
        text.count
    }

    var canPost: Bool {
        let hasText = !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        let hasMedia = !selectedImages.isEmpty
        let withinLimit = text.count <= 300
        return (hasText || hasMedia) && withinLimit
    }

    var canSaveDraft: Bool {
        !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || !selectedImages.isEmpty
    }

    init(replyTo: ReplyRef? = nil, draft: PostDraft? = nil) {
        self.replyTo = replyTo
        self.selectedLanguage = LanguagePreferences.shared.preferredLanguage
        self.moderationSettings = .default

        // Load draft if provided
        if let draft = draft {
            self.currentDraftId = draft.id
            self.text = draft.text
            self.imageAltTexts = draft.imageAltTexts
            self.selectedLanguage = Language.allLanguages.first { $0.id == draft.languageId } ?? LanguagePreferences.shared.preferredLanguage
            self.moderationSettings = draft.moderationSettings

            // Convert image data back to PlatformImages
            self.selectedImages = draft.imageData.compactMap { PlatformImage(data: $0) }
        }
    }

    func loadSelectedPhotos() async {
        selectedImages.removeAll()
        imageAltTexts.removeAll()

        for item in selectedPhotoItems {
            if let data = try? await item.loadTransferable(type: Data.self),
               let image = PlatformImage(data: data) {
                selectedImages.append(image)
                imageAltTexts.append("") // Initialize with empty alt text
            }
        }
    }

    func removeImage(at index: Int) {
        guard index < selectedImages.count else { return }
        selectedImages.remove(at: index)
        selectedPhotoItems.remove(at: index)
        if index < imageAltTexts.count {
            imageAltTexts.remove(at: index)
        }
    }

    func updateAltText(at index: Int, text: String) {
        guard index < imageAltTexts.count else { return }
        imageAltTexts[index] = text
    }

    // MARK: - Mention Autocomplete

    func detectMention() {
        // Cancel any ongoing search
        mentionSearchTask?.cancel()

        // Find the current word being typed
        let cursorIndex = min(cursorPosition, text.count)
        let textBeforeCursor = String(text.prefix(cursorIndex))

        // Find the last @ symbol before cursor
        if let lastAtIndex = textBeforeCursor.lastIndex(of: "@") {
            // Get text from @ to cursor
            let mentionText = String(textBeforeCursor.suffix(from: lastAtIndex).dropFirst()) // Remove @

            // Check if there's a space (which would end the mention)
            if !mentionText.contains(" ") && !mentionText.isEmpty {
                currentMentionQuery = mentionText
                searchMentions(query: mentionText)
            } else if mentionText.isEmpty {
                // User just typed @, show popular suggestions
                currentMentionQuery = ""
                searchMentions(query: "")
            } else {
                // Space found, clear suggestions
                clearMentionSuggestions()
            }
        } else {
            // No @ found, clear suggestions
            clearMentionSuggestions()
        }
    }

    private func searchMentions(query: String) {
        // Cancel previous search
        mentionSearchTask?.cancel()

        isSearchingMentions = true

        mentionSearchTask = Task {
            do {
                // Add a small delay to avoid too many API calls
                try await Task.sleep(nanoseconds: 300_000_000) // 0.3 seconds

                guard !Task.isCancelled else { return }

                var results: [Profile] = []
                var seenDIDs = Set<String>()

                // Get current user's DID
                guard let currentUserDID = client.session?.did else {
                    isSearchingMentions = false
                    return
                }

                // 1. Search follows (people you follow)
                do {
                    let follows = try await client.getFollows(actor: currentUserDID, limit: 50)
                    guard !Task.isCancelled else { return }

                    for profile in follows {
                        // Filter by query
                        if query.isEmpty ||
                           profile.handle.localizedCaseInsensitiveContains(query) ||
                           (profile.displayName?.localizedCaseInsensitiveContains(query) ?? false) {

                            if !seenDIDs.contains(profile.did) {
                                results.append(profile)
                                seenDIDs.insert(profile.did)
                            }
                        }
                    }
                    print("📍 Found \(results.count) matches in follows for '\(query)'")
                } catch {
                    print("⚠️ Failed to search follows: \(error)")
                }

                guard !Task.isCancelled else { return }

                // 2. Search followers (people who follow you)
                do {
                    let followers = try await client.getFollowers(actor: currentUserDID, limit: 50)
                    guard !Task.isCancelled else { return }

                    for profile in followers {
                        // Filter by query
                        if query.isEmpty ||
                           profile.handle.localizedCaseInsensitiveContains(query) ||
                           (profile.displayName?.localizedCaseInsensitiveContains(query) ?? false) {

                            if !seenDIDs.contains(profile.did) {
                                results.append(profile)
                                seenDIDs.insert(profile.did)
                            }
                        }
                    }
                    print("📍 Found \(results.count) total matches (including followers) for '\(query)'")
                } catch {
                    print("⚠️ Failed to search followers: \(error)")
                }

                guard !Task.isCancelled else { return }

                // 3. Search all users if we need more results
                if results.count < 10 {
                    do {
                        let searchResults = try await client.searchUsers(query: query, limit: 10)
                        guard !Task.isCancelled else { return }

                        for profile in searchResults {
                            if !seenDIDs.contains(profile.did) {
                                results.append(profile)
                                seenDIDs.insert(profile.did)
                            }
                        }
                        print("📍 Found \(results.count) total matches (including general search) for '\(query)'")
                    } catch {
                        print("⚠️ Failed to search users: \(error)")
                    }
                }

                guard !Task.isCancelled else { return }

                // Limit to 10 results
                mentionSuggestions = Array(results.prefix(10))
                isSearchingMentions = false
                print("✅ Final mention suggestions: \(mentionSuggestions.count) for '\(query)'")
            } catch {
                if !Task.isCancelled {
                    print("Failed to search mentions: \(error)")
                    isSearchingMentions = false
                }
            }
        }
    }

    func insertMention(_ profile: Profile) {
        let cursorIndex = min(cursorPosition, text.count)
        let textBeforeCursor = String(text.prefix(cursorIndex))

        // Find the last @ symbol
        if let lastAtIndex = textBeforeCursor.lastIndex(of: "@") {
            let beforeAt = String(text.prefix(upTo: lastAtIndex))
            let afterCursor = String(text.suffix(from: text.index(text.startIndex, offsetBy: cursorIndex)))

            // Insert the mention
            let mention = "@\(profile.handle) "
            text = beforeAt + mention + afterCursor

            // Update cursor position to after the inserted mention
            cursorPosition = beforeAt.count + mention.count

            // Clear suggestions
            clearMentionSuggestions()
        }
    }

    func clearMentionSuggestions() {
        currentMentionQuery = nil
        mentionSuggestions = []
        isSearchingMentions = false
        mentionSearchTask?.cancel()
    }

    // MARK: - Hashtag Autocomplete

    func detectHashtag() {
        // Cancel any ongoing search
        hashtagSearchTask?.cancel()

        // Find the current word being typed
        let cursorIndex = min(cursorPosition, text.count)
        let textBeforeCursor = String(text.prefix(cursorIndex))

        // Find the last # symbol before cursor
        if let lastHashIndex = textBeforeCursor.lastIndex(of: "#") {
            // Get text from # to cursor
            let hashtagText = String(textBeforeCursor.suffix(from: lastHashIndex).dropFirst()) // Remove #

            // Check if there's a space (which would end the hashtag)
            if !hashtagText.contains(" ") && !hashtagText.isEmpty {
                currentHashtagQuery = hashtagText
                searchHashtags(query: hashtagText)
            } else if hashtagText.isEmpty {
                // User just typed #, show popular hashtags
                currentHashtagQuery = ""
                searchHashtags(query: "")
            } else {
                // Space found, clear suggestions
                clearHashtagSuggestions()
            }
        } else {
            // No # found, clear suggestions
            clearHashtagSuggestions()
        }
    }

    private func searchHashtags(query: String) {
        // Cancel previous search
        hashtagSearchTask?.cancel()

        isSearchingHashtags = true

        hashtagSearchTask = Task {
            do {
                // Add a small delay to avoid too many API calls
                try await Task.sleep(nanoseconds: 300_000_000) // 0.3 seconds

                guard !Task.isCancelled else { return }

                // Search for posts with this hashtag
                let searchQuery = query.isEmpty ? "popular" : "#\(query)"
                let response = try await client.searchPosts(query: searchQuery, limit: 20)

                guard !Task.isCancelled else { return }

                // Extract unique hashtags from posts
                var hashtags = Set<String>()
                for post in response.posts {
                    if let tags = post.record.tags {
                        for tag in tags {
                            hashtags.insert(tag)
                        }
                    }
                }

                // Convert to sorted array and filter by query
                let filteredHashtags = hashtags.sorted().filter { tag in
                    query.isEmpty || tag.lowercased().hasPrefix(query.lowercased())
                }

                hashtagSuggestions = Array(filteredHashtags.prefix(10))
                isSearchingHashtags = false
                print("✅ Found \(hashtagSuggestions.count) hashtag suggestions for '\(query)'")
            } catch {
                if !Task.isCancelled {
                    print("Failed to search hashtags: \(error)")
                    isSearchingHashtags = false
                }
            }
        }
    }

    func insertHashtag(_ tag: String) {
        let cursorIndex = min(cursorPosition, text.count)
        let textBeforeCursor = String(text.prefix(cursorIndex))

        // Find the last # symbol
        if let lastHashIndex = textBeforeCursor.lastIndex(of: "#") {
            let beforeHash = String(text.prefix(upTo: lastHashIndex))
            let afterCursor = String(text.suffix(from: text.index(text.startIndex, offsetBy: cursorIndex)))

            // Insert the hashtag
            let hashtag = "#\(tag) "
            text = beforeHash + hashtag + afterCursor

            // Update cursor position to after the inserted hashtag
            cursorPosition = beforeHash.count + hashtag.count

            // Clear suggestions
            clearHashtagSuggestions()
        }
    }

    func clearHashtagSuggestions() {
        currentHashtagQuery = nil
        hashtagSuggestions = []
        isSearchingHashtags = false
        hashtagSearchTask?.cancel()
    }

    func post() async -> Bool {
        guard canPost else { return false }

        isPosting = true
        errorMessage = nil

        do {
            // Upload images if any are selected
            var uploadedImages: [UploadedImage] = []
            for (index, image) in selectedImages.enumerated() {
                let altText = index < imageAltTexts.count ? imageAltTexts[index] : nil
                let finalAltText = (altText?.isEmpty == false) ? altText : nil
                if let uploaded = try await client.uploadImage(image, altText: finalAltText) {
                    uploadedImages.append(uploaded)
                }
            }

            _ = try await client.createPost(text: text, reply: replyTo, images: uploadedImages.isEmpty ? nil : uploadedImages, langs: [selectedLanguage.id], moderationSettings: moderationSettings)

            // Track post creation
            Analytics.logEvent("post_created", parameters: [
                "post_type": replyTo != nil ? "reply" : "original",
                "character_count": text.count,
                "has_media": !selectedImages.isEmpty,
                "media_count": selectedImages.count
            ])
            print("📊 Analytics: Logged post_created (\(replyTo != nil ? "reply" : "original"), \(selectedImages.count) images)")

            // Delete draft if this was from a draft
            if let draftId = currentDraftId {
                if let draft = DraftManager.shared.getDraft(id: draftId) {
                    DraftManager.shared.deleteDraft(draft)
                }
            }

            text = ""
            selectedImages = []
            selectedPhotoItems = []
            isPosting = false
            return true
        } catch let error as ATProtoError {
            // Use the detailed error message from ATProtoError
            errorMessage = error.localizedDescription
            isPosting = false
            print("❌ Post failed: \(error)")
            return false
        } catch {
            // Generic error - add PDS context
            let pdsURL = client.session?.pdsURL ?? "unknown"
            errorMessage = "Post failed: \(error.localizedDescription)\n\nPDS: \(pdsURL)"
            isPosting = false
            print("❌ Post failed with generic error: \(error)")
            return false
        }
    }

    func saveDraft() {
        guard canSaveDraft else { return }

        // Convert images to data
        let imageData = selectedImages.compactMap { $0.jpegData(compressionQuality: 0.8) }

        let draft = PostDraft(
            id: currentDraftId ?? UUID(),
            text: text,
            imageData: imageData,
            imageAltTexts: imageAltTexts,
            languageId: selectedLanguage.id,
            moderationSettings: moderationSettings
        )

        DraftManager.shared.saveDraft(draft)
        print("💾 Draft saved: \(draft.id)")
    }

    func deleteDraft() {
        if let draftId = currentDraftId,
           let draft = DraftManager.shared.getDraft(id: draftId) {
            DraftManager.shared.deleteDraft(draft)
            print("🗑️ Draft deleted: \(draftId)")
        }
    }
}

// MARK: - Uploaded Image Model
struct UploadedImage {
    let blob: BlobRef
    let aspectRatio: AspectRatio?
    let alt: String?

    struct AspectRatio {
        let width: Int
        let height: Int
    }
}

struct BlobRef: Codable {
    let type: String
    let ref: BlobLink
    let mimeType: String
    let size: Int

    enum CodingKeys: String, CodingKey {
        case type = "$type"
        case ref
        case mimeType
        case size
    }
}

struct BlobLink: Codable {
    let link: String

    enum CodingKeys: String, CodingKey {
        case link = "$link"
    }
}
